src/components/bookmark-button/with-state.tsx

```diff
@@ -0,0 +1,60 @@
+import React, { useCallback, useState } from 'react'
+import Dialog from 'components/dialog'
+import useAuthentication from 'hooks/use-authentication'
+import { BookmarkButtonProps } from './types'
+import BookmarkSignInPrompt from './sign-in-dialog'
+
+/**
+ * This HOC serves as a 'controller'
+ * for rendering the various state of bookmark
+ * interaction. It accepts a UI component to render,
+ * will post / delete based on interactions and
+ * handle the dialog if unauthenticed & toast notifications
+ */
+
+export function withState(BookmarkComponent: React.FC<BookmarkButtonProps>) {
```

Comment:
Could use a better name here, open to suggestions.

Reply:
I wonder if I should refactor this back to `withDialog` for now as we are still sussing out how to handle the full data integration 🤔

Reply:
I'm on board with either name, since it feels like everything's nicely document, but agreed that `withDialog` does seem a little more to-the-point with the current scope 👍